File: Leonidas/Final_Project/files/bootstrap.py
# crypto_project/bootstrap.py

import random
from .crypto_utils import are_relatively_prime, miller_rabin_test


INIT_NOT_PRIME = 15
ROUNDS = 20


class Bootstrap:
    """
    Uses Python's random module to generate initial values and candidates.
    """

    def __init__(self):
        # SystemRandom uses OS randomness
        self.rng = random.SystemRandom()

    def random_number(self, bit_length):
        if bit_length <= 0:
            raise ValueError("bit_length must be > 0")

        result = 0
        bytes_needed = (bit_length + 7) // 8

        for _ in range(bytes_needed):
            random_byte = self.rng.randrange(0, 256)
            result = (result << 8) | random_byte

        # Mask to desired bit length
        mask = (1 << bit_length) - 1
        result = result & mask

        # Set the highest bit to ensure it really has bit_length bits
        result |= (1 << (bit_length - 1))

        return result

    def random_odd_number(self, bit_length):
        number = self.random_number(bit_length)
        if number % 2 == 0:
            number += 1
        return number

    def random_in_range(self, min_value, max_value):
        if min_value > max_value:
            raise ValueError("min_value cannot be greater than max_value")

        min_value = int(min_value)
        max_value = int(max_value)
        range_val = max_value - min_value + 1
        bits_needed = range_val.bit_length()
        result = range_val

        while result >= range_val:
            candidate = self.random_number(bits_needed)
            result = candidate

        return min_value + result

    def generate_seed(self, n):
        """
        Generates a seed s with 2 <= s <= n-1 and gcd(s, n) = 1.
        """
        n = int(n)
        seed = self.random_in_range(2, n - 1)
        while not are_relatively_prime(seed, n):
            seed = self.random_in_range(2, n - 1)
        return seed

    def generate_prime_congruent_3_mod_4(self, bit_length):
        """
        Generates a prime number p with p ≡ 3 (mod 4).
        """
        candidate = INIT_NOT_PRIME
        while not miller_rabin_test(candidate, ROUNDS, self):
            candidate = self.random_odd_number(bit_length)
            if candidate % 4 != 3:
                candidate += 2
        return candidate

    def generate_prime_in_range(self, min_value, max_value):
        """
        Generates a prime number in the range [min_value, max_value].
        """
        candidate = INIT_NOT_PRIME
        while not miller_rabin_test(candidate, ROUNDS, self):
            candidate = self.random_in_range(min_value, max_value)
            if candidate % 2 == 0:
                candidate += 1
        return candidate

    def generate_prime_in_range_congruent_3_mod_4(self, min_value, max_value):
        """
        Generates a prime number in the range [min, max] with p ≡ 3 (mod 4).
        """
        candidate = INIT_NOT_PRIME
        while not miller_rabin_test(candidate, ROUNDS, self):
            candidate = self.random_in_range(min_value, max_value)
            if candidate % 2 == 0:
                candidate += 1
            if candidate % 4 != 3:
                candidate += 2
        return candidate

File: Alexan/final_project/src/Bootstrap.cpp
#include "Bootstrap.h"
#include "crypto_utils.h"

const int INIT_NOT_PRIME = 15;
const int ROUNDS = 20;

Bootstrap::Bootstrap() {
    std::random_device rd;
    rng.seed(rd());
}

mpz_class Bootstrap::randomNumber(size_t bitLength)
{
    mpz_class result = 0;
    size_t bytesNeeded = (bitLength + 7) / 8;
    
    for (size_t i = 0; i < bytesNeeded; i++) {
        uint8_t randomByte = static_cast<uint8_t>(rng() & 0xFF);
        mpz_class byteValue = randomByte;
        byteValue = byteValue << (8 * i);
        result += byteValue;
    }
    
    // Mask to exact bit length
    mpz_class mask = (mpz_class(1) << bitLength) - 1;
    result &= mask;  // Keep only the lower 256 bits
    
    return result;
}

mpz_class Bootstrap::randomOddNumber(size_t bitLength)
{
    mpz_class number = randomNumber(bitLength);
    
    // Set high bit to guarantee a number with bitLength bits.
    mpz_setbit(number.get_mpz_t(), bitLength - 1);
    
    if (number % 2 == 0) {
        number += 1;  // Make it odd.
    }
    
    return number;
}

mpz_class Bootstrap::randomInRange(const mpz_class& min, const mpz_class& max)
{
    mpz_class range = max - min + 1;
    size_t bits_needed = mpz_sizeinbase(range.get_mpz_t(), 2);
    
    mpz_class result = range;  // Initialize to invalid value
    
    while (result >= range) {
        result = randomNumber(bits_needed);
    }
    
    return min + result;
}

mpz_class Bootstrap::generateSeed(const mpz_class& n)
{
    mpz_class seed = randomInRange(2, n - 1);

    while (!areRelativelyPrime(n, seed)) {
        seed = randomInRange(2, n - 1);
    }

    return seed;
}

mpz_class Bootstrap::generatePrimeCongruent3Mod4(size_t bitLength)
{
    mpz_class candidate = INIT_NOT_PRIME;  // Thus starting our while loop.
    
    while (!millerRabinTest(candidate, ROUNDS, *this)) {
        candidate = randomOddNumber(bitLength);
        
        // Ensure it's ≡ 3 (mod 4)
        // Odd numbers are either ≡ 1 (mod 4) or ≡ 3 (mod 4)
        if (candidate % 4 != 3) {
            candidate += 2;  // Change from ≡ 1 to ≡ 3 (mod 4)
        }
    }
        
    return candidate;
}

mpz_class Bootstrap::generatePrimeInRange(const mpz_class& min, const mpz_class& max)
{
    mpz_class candidate = INIT_NOT_PRIME;
    
    while (!millerRabinTest(candidate, ROUNDS, *this)) {
        candidate = randomInRange(min, max);
        
        // Make it odd (all primes except 2 are odd)
        if (candidate % 2 == 0) {
            candidate += 1;
        }
    }
    
    return candidate;
}

mpz_class Bootstrap::generatePrimeInRangeCongruent3Mod4(const mpz_class& min, const mpz_class& max)
{
    mpz_class candidate = INIT_NOT_PRIME;
    
    while (!millerRabinTest(candidate, ROUNDS, *this)) {
        candidate = randomInRange(min, max);
        
        if (candidate % 2 == 0) {
            candidate += 1;  // Make it odd.
        }
        
        // Ensure it's ≡ 3 (mod 4)
        if (candidate % 4 != 3) {
            candidate += 2;  // Change from ≡ 1 to ≡ 3 (mod 4)
        }
    }
    
    return candidate;
}